src/Servers/Kestrel/Core/src/KestrelConfigurationLoader.cs

```diff
@@ -435,14 +436,124 @@ private X509Certificate2 LoadCertificate(CertificateConfig certInfo, string endp
             {
-                var env = Options.ApplicationServices.GetRequiredService<IHostEnvironment>();
-                return new X509Certificate2(Path.Combine(env.ContentRootPath, certInfo.Path), certInfo.Password);
+                var environment = Options.ApplicationServices.GetRequiredService<IHostEnvironment>();
+                var certificatePath = Path.Combine(environment.ContentRootPath, certInfo.Path);
+                if (certInfo.KeyPath != null)
+                {
+                    var certificateKeyPath = Path.Combine(environment.ContentRootPath, certInfo.KeyPath);
+                    var certificate = GetCertificate(certInfo, certificatePath, certificateKeyPath);
+
+                    if (certificate != null)
+                    {
+                        certificate = LoadCertificateKey(certificate, certificateKeyPath, certInfo.Password);
+
+                        if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+                        {
+                            return PersistKey(certificate);
+                        }
+
+                        return certificate;
+                    }
+
+                    throw new InvalidOperationException(CoreStrings.InvalidPemKey);
+                }
+
+                return new X509Certificate2(Path.Combine(environment.ContentRootPath, certInfo.Path), certInfo.Password);
             }
             else if (certInfo.IsStoreCert)
             {
                 return LoadFromStoreCert(certInfo);
             }
             return null;
+
+            static X509Certificate2 PersistKey(X509Certificate2 fullCertificate)
+            {
+                // We need to force the key to be persisted.
+                // See https://github.com/dotnet/runtime/issues/23749
+                var certificateBytes = fullCertificate.Export(X509ContentType.Pkcs12, "");
+                return new X509Certificate2(certificateBytes, "", X509KeyStorageFlags.DefaultKeySet);
+            }
+
+            static X509Certificate2 LoadCertificateKey(X509Certificate2 certificate, string keyPath, string password)
+            {
+                var keyText = File.ReadAllText(keyPath);
+                return TryAttachPemRSAKey(certificate, keyText, password) ??
+                    TryAttachPemDSAKey(certificate, keyText, password) ??
+                    TryAttachPemECDSAKey(certificate, keyText, password);
+            }
+
+            static X509Certificate2 GetCertificate(CertificateConfig certInfo, string certificatePath, string certificateKeyPath)
+            {
+                if (X509Certificate2.GetCertContentType(certificatePath) == X509ContentType.Cert)
+                {
+                    return new X509Certificate2(certificatePath);
+                }
+
+                return null;
+            }
+        }
+
+        private static X509Certificate2 TryAttachPemRSAKey(X509Certificate2 certificate, string keyText, string password)
```

Comment:
The TryAttach methods look like they do the same thing, but with different algorithms.

You could remove some duplication here by resolving an `AsymmetricAlgorithm` from the oid, and then having a single method that loads the cert using the given algorithm.

Reply:
Nit: "Try" implies a boolean return value. Something like GetCertWithAttachedKeyOrNull is a little better.

Reply:
I've refactor the method to make the decision in the caller and just handle the loading here.